src/udev/udev-watch.c

```diff
@@ -10,16 +10,97 @@
 #include "fs-util.h"
+#include "mkdir.h"
 #include "parse-util.h"
-#include "random-util.h"
+#include "path-util.h"
+#include "rm-rf.h"
+#include "stdio-util.h"
+#include "string-util.h"
 #include "udev-watch.h"
 
-#define SAVE_WATCH_HANDLE_MAX_RETRIES  128
-#define MAX_RANDOM_DELAY (100 * USEC_PER_MSEC)
-#define MIN_RANDOM_DELAY ( 10 * USEC_PER_MSEC)
+int device_new_from_watch_handle(sd_device **ret, int wd) {
+        _cleanup_closedir_ DIR *dir = NULL;
+        char wd_str[DECIMAL_STR_MAX(int)];
+
+        assert(ret);
+
+        /* At the time this function is called, there exist the following possibilities:
+         * A. No matching symlink exists. This happens (A-1) when the watch handle is assigned to a
+         *    device, but it has not been saved to a symlink yet. Or (A-2) the device is already
+         *    removed and the corresponding 'remove' uevent is already processed or being processed.
+         * B. A wrong symlink exists. This happens when a device, which the watch handle was previously
+         *    assigned to, has been already removed, but the corresponding 'remove' uevent has not been
+         *    processed yet. And the watch handle has been already assigned to another device.
+         *
+         * So, in the worst case, there exist multiple symlinks for each watch handle. However, even in
+         * such case, for each device handle, there exists at most one symlink from the device ID
+         * corresponds to a *existing* device. See also comments in udev_watch_end(). Hence, in the
+         * below, it is OK to return the first device obtained by sd_device_new_from_device_id().
+         *
+         * Of course, there exists another worst case that there exists no symlink to point the
+         * requested watch handle. That is, the case A in the above. But, the case A-2 does not cause
+         * any issue, as the device does not exist anymore. Unfortunately, the case A-1 cannot be
+         * salvaged. But the timespan of A-1 should be small enough. Moreover, when we process a block
+         * device, we lock its backing whole block device (except for some special kind of devices such
+         * as MD devices). Hence, applications which follow our advisory
+         * https://systemd.io/BLOCK_DEVICE_LOCKING do not trigger such inotify events. And, usually,
+         * no character device is watched. */
+
+        if (wd < 0)
+                return -EBADF;
+
+        xsprintf(wd_str, "%d", wd);
+
+        dir = opendir("/run/udev/watch");
+        if (!dir)
+                return -errno;
+
+        FOREACH_DIRENT_ALL(de, dir, break) {
+                _cleanup_(sd_device_unrefp) sd_device *dev = NULL;
+                _cleanup_free_ char *buf = NULL;
+
+                if (dot_or_dot_dot(de->d_name))
+                        continue;
+
+                if (readlinkat_malloc(dirfd(dir), de->d_name, &buf) < 0)
+                        continue;
+
+                if (!streq(buf, wd_str))
```

Comment:
i think it would be safer to parse the specified buffer via `safe_atoi()` and then do a numeric comparison wiht `ẁd`, instead of first formatting `wd` into `wd_str`.